src/langchain_google_spanner/chat_message_history.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Cloud Spanner-based chat message history"""
+from __future__ import annotations
+
+import uuid
+from typing import List, Optional
+
+from google.cloud import spanner
+from google.cloud.spanner_admin_database_v1.types import DatabaseDialect  # type: ignore
+from google.cloud.spanner_v1 import param_types
+from google.cloud.spanner_v1.data_types import JsonObject
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import BaseMessage, message_to_dict, messages_from_dict
+
+OPERATION_TIMEOUT_SECONDS = 240
+
+COLUMN_FAMILY = "langchain"
+COLUMN_NAME = "history"
+
+
+class SpannerChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history that stores history in Spanner.
+
+    Args:
+        instance_id: The Spanner instance to use for chat message history.
+        database_id: The Spanner database to use for chat message history.
+        table_name: The Spanner table to use for chat message history.
+        session_id: Optional. The existing session ID.
+    """
+
+    def __init__(
+        self,
+        instance_id: str,
+        database_id: str,
+        table_name: str = "message_store",
+        session_id: Optional[str] = None,
+        client: Optional[spanner.Client] = None,
+    ) -> None:
+        self.instance_id = instance_id
+        self.database_id = database_id
+        self.table_name = table_name
+        self.client = client or spanner.Client()
+        self.instance = self.client.instance(instance_id)
+        if not self.instance.exists():
+            raise Exception("Instance doesn't exist.")
+        self.database = self.instance.database(database_id)
+        if not self.database.exists():
+            raise Exception("Database doesn't exist.")
+        self.database.reload()
+        self.dialect = self.database.database_dialect
+        self.session_id = session_id or uuid.uuid4().hex
+        self._create_table_if_not_exists()
```

Comment:
Kurtis and I have decided that table creation should happen explicitly outside the table, see convo https://github.com/googleapis/langchain-google-cloud-sql-mssql-python/pull/9#discussion_r1478561574.